### DISCUSSION TITLE
Is it possible to redirect route with async code?

### DISCUSSION THREAD

**Discussion by MoonyDevs** [jonataslaw/getx#1566] [Q&A]
Hi all, does anybody has used https://github.com/jonataslaw/getx#redirect redirect with async code? 
For instance if authservice.isAuthenticated return a future like this: 

class AuthMiddleware extends GetMiddleware {
  @override
  RouteSettings? redirect(String? route) async {
    final AuthenticationApi _authService = AuthenticationApi();
    bool isUserAuthenticated = await _authService.isUserAuthenticated();
    return RouteSettings(
        name: isUserAuthenticated ? route : Navigation.login);
  }
}

**Comment by manikantag** [comment]
Hi, using `redirectDelegate()`, you can do that. Remember to use `GetMaterialApp.router()` to run the app.

See example app: https://github.com/jonataslaw/getx/blob/master/example_nav2/lib/app/middleware/auth_middleware.dart

**Reply by tony123S** [reply]
@manikantag  hye, I tried your answer, but the method not calling. I need to check authToken in splashScreen. If not null and empty, redirect to mainPage, otherwise redirect to login page.

**main**
```

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return GetMaterialApp(
      initialRoute: '/splashScreen',
      getPages: [
        GetPage(
            name: '/splashScreen',
            page: () => SplashScreen(),
            binding: Bind(),
            middlewares: [GlobalMiddleware()]),
        GetPage(
            name: '/login',
            page: () => LoginPage(),
            binding: Bind(),
           ),
        GetPage(
            name: '/mainPage', page: () => MainPage(), binding: Bind())
      ],
          ....
        ),
      ),
      // home: SplashScreen(),
    );
  }
}
```

**GlobalMiddleware**

```

class GlobalMiddleware extends GetMiddleware {
  @override
  Future<GetNavConfig?> redirectDelegate(GetNavConfig route) async {
    print("calling here");
    var user = await MySharedPreferences().getUser();

    if (user.auth_token != "") {
      print("fefe");
      return GetNavConfig.fromRoute("/mainPage");
    } else if (user.auth_token == "" && user.auth_token == null) {
      print("efef");
      return GetNavConfig.fromRoute("/login");
    }
    print("efecsf");
    return await super.redirectDelegate(route);
  }
}

```

**Output**

```
[GETX] Instance "GetMaterialController" has been created
[GETX] Instance "GetMaterialController" has been initialized
[GETX] Redirect to null
[GETX] GOING TO ROUTE /splashScreen
I/us_work_factor( 3461): ProcessProfilingInfo new_methods=584 is saved saved_to_disk=1 resolve_classes_delay=8000
V/MiuiFrameworkFactory( 3461): get AllImpl object = android.common.MiuiFrameworkFactoryImpl@e330ee4
W/MirrorManager( 3461): this model don't Support

```

Do you know why? Thanks.

**Reply by manikantag** [reply]
Hey, as I mentioned you need to return `GetMaterialApp.router()` from your `MyApp.build()`

**Reply by manikantag** [reply]
Currently you are returning `GetMaterialApp`

**Reply by tony123S** [reply]
> Hey, as I mentioned you need to return `GetMaterialApp.router()` from your `MyApp.build()`
```

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return GetMaterialApp.router(
      // initialRoute: '/splashScreen',
      getPages: [
        GetPage(
            name: '/splashScreen',
            page: () => SplashScreen(),
            binding: Bind(),
            middlewares: [GlobalMiddleware()]),
        GetPage(
          name: '/login',
          page: () => LoginPage(),
          binding: Bind(),
        ),
        GetPage(
            name: '/mainPage', page: () => MainPage(), binding: Bind())
      ],
            ...
    );
  }
}
```

After change to `GetMaterialApp.router()`, it able to redirect to login page. But when I press submit button in login page to go to main page,I get another error.

` Get.toNamed('/mainPage');`  // after login button is pressed

**Error**
```

Reloaded 1 of 1342 libraries in 892ms.
E/flutter ( 7696): [ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: Navigator.onGenerateRoute was null, but the route named "/mainPage" was referenced.
E/flutter ( 7696): To use the Navigator API with named routes (pushNamed, pushReplacementNamed, or pushNamedAndRemoveUntil), the Navigator must be provided with an onGenerateRoute handler.
E/flutter ( 7696): The Navigator was:
```

**Comment by manikantag** [comment]
Use `Get.rootDelegate.toXXX()` for routing.

**Reply by tony123S** [reply]
Thanks ! But how about `Get.offAll`?

I get error on this line

`Get.offAll(LoginPage(), binding: Bind());`

**Error**
`
E/flutter (18076): [ERROR:flutter/shell/common/shell.cc(93)] Dart Unhandled Exception: 'package:flutter/src/widgets/overlay.dart': Failed assertion: line 423 pos 7: 'newEntriesList.every((OverlayEntry entry) => entry._overlay == null || entry._overlay == this)': One or more of the specified entries are already present in another Overlay., stack trace: #0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:46:39)`

**Reply by fmdogan** [reply]
Late answer but may be helpful for newbies..

You should be aware of that with GetMaterialApp.router(..), you are using Navigator 2..
Get has another api to use Navigator 2 which is rootDelegate.
if you use Get.toNamed(..); with Navigator 2 api, you will need to use Get.rootDelegate.toNamed(..)..

So; routing api's on Get is for Navigator 1, api's on Get.rootDelegate are for Navigator 2.
